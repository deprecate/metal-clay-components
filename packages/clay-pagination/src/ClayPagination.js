import 'clay-link';
// eslint-disable-next-line
import { ClayDropdownBase } from 'clay-dropdown';
import {Config} from 'metal-state';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import templates from './ClayPagination.soy.js';

/**
 * Metal ClayPagination component.
 */
class ClayPagination extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayPagination.STATE = {
	/**
	 * Href to create each page complete href.
	 * @instance
	 * @memberof ClayPagination
	 * @type {?string|undefined}
	 * @default undefined
	 */
	baseHref: Config.string(),

	/**
	 * Current page number.
	 * @instance
	 * @memberof ClayPagination
	 * @type {!number}
	 * @default undefined
	 */
	currentPage: Config.number().required(),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayPagination
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayPagination
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayPagination
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * Total number of pages to show.
	 * @instance
	 * @memberof ClayPagination
	 * @type {!number}
	 * @default undefined
	 */
	totalPages: Config.number().required(),
};

defineWebComponent('clay-pagination', ClayPagination);

Soy.register(ClayPagination, templates);

export {ClayPagination};
export default ClayPagination;
