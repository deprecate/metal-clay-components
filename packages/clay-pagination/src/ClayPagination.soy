{namespace ClayPagination}

/**
 * The renders a pagination.
 */
{template .render}
	{@param spritemap: string}
	{@param total: number}
	{@param? _handleClickItem: any}
	{@param? _handleClickNext: any}
	{@param? _handleClickPrev: any}
	{@param? _itemsEllipse: ?}
	{@param? baseHref: string}
	{@param? ellipseRange: number}
	{@param? offset: number}
	{@param? page: number}

	<ul class="pagination">
		{call .item}
			{param disabled: $page == $offset ? true : false /}
			{param elementClasses: 'page-link' /}
			{param event: $_handleClickPrev /}
			{param spritemap: $spritemap /}
			{param symbol: 'angle-left' /}
		{/call}

		{call .range}
			{param _handleClickItem: $_handleClickItem /}
			{param _itemsEllipse: $_itemsEllipse /}
			{param baseHref: $baseHref /}
			{param ellipseRange: $ellipseRange /}
			{param offset: $offset /}
			{param page: $page /}
			{param total: $total /}
		{/call}

		{call .item}
			{param disabled: $page == $total ? true : false /}
			{param elementClasses: 'page-link' /}
			{param event: $_handleClickNext /}
			{param spritemap: $spritemap /}
			{param symbol: 'angle-right' /}
		{/call}
	</ul>
{/template}

/**
 * Renders a list of items the pages.
 */
{template .range}
	{@param total: number}
	{@param? _handleClickItem: any}
	{@param? _itemsEllipse: ?}
	{@param? baseHref: string}
	{@param? ellipseRange: number}
	{@param? offset: number}
	{@param? page: number}

	{for $index in range($total - $offset + 1)}
		{let $active: $page == $index + $offset ? true : false /}

		{if $total > $ellipseRange and $index < ($ellipseRange - 3)}
			{call .item}
				{param active: $active /}
				{param event: $_handleClickItem /}
				{param href: $baseHref + ($index + $offset) /}
				{param label: '' + ($index + $offset) /}
			{/call}
		{elseif $total <= $ellipseRange}
			{call .item}
				{param active: $active /}
				{param event: $_handleClickItem /}
				{param href: $baseHref + ($index + $offset) /}
				{param label: '' + ($index + $offset) /}
			{/call}
		{/if}
	{/for}

	{if $total > $ellipseRange}
		<li class="page-item">
			{call ClayDropdownBase.render}
				{param items: $_itemsEllipse /}
				{param triggerElementClasses: 'page-link' /}
				{param triggerLabel: '...' /}
				{param triggerStyle: 'link' /}
			{/call}
		</li>

		{let $active: $page == $total ? true : false /}

		{call .item}
			{param active: $active /}
			{param event: $_handleClickItem /}
			{param href: $baseHref + $total /}
			{param label: '' + $total /}
		{/call}
	{/if}
{/template}

/**
 * Renders a item in page.
 */
{template .item}
	{@param? active: bool}
	{@param? disabled: bool}
	{@param? event: any}
	{@param? href: string}
	{@param? label: string}
	{@param? spritemap: string}
	{@param? symbol: string}

	{let $classes kind="text"}
		page-item
		{if $active}
			{sp}active
		{/if}

		{if $disabled}
			{sp}disabled
		{/if}
	{/let}

	<li class="{$classes}">
		{if $spritemap and $symbol}
			{call ClayLink.render}
				{param elementClasses: 'page-link' /}
				{param events: ['click': $event] /}
				{param icon: [
					'alignment': 'left',
					'monospaced': false,
					'spritemap': $spritemap,
					'symbol': $symbol
				] /}
			{/call}
		{else}
			{call ClayLink.render}
				{param elementClasses: 'page-link' /}
				{param events: ['click': $event] /}
				{param href: $href /}
				{param label: $label /}
			{/call}
		{/if}
	</li>
{/template}
