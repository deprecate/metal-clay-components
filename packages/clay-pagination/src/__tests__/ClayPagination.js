import ClayPagination from '../ClayPagination.js';

let clayPagination;
let icon = 'icons.svg';

describe('ClayPagination', function() {
  afterEach(() => {
    if (clayPagination) {
      clayPagination.dispose();
    }
  });

  it('should render a pagination markup', () => {
    clayPagination = new ClayPagination({
      spritemap: icon,
      total: 1,
    });

    expect(clayPagination).toMatchSnapshot();
  });

  it('should render a pagination with total of the pages visible', () => {
    clayPagination = new ClayPagination({
      spritemap: icon,
      total: 4,
    });

    expect(clayPagination).toMatchSnapshot();
  });

  // eslint-disable-next-line
  it('should render a pagination with 4 pages visible and 3 pages in dropdown', () => {
    clayPagination = new ClayPagination({
      spritemap: icon,
      total: 7,
    });

    expect(clayPagination).toMatchSnapshot();
  });

  // eslint-disable-next-line
  it('should render a pagination with 4 pages visible and 3 pages in dropdown with page active in dropdown', () => {
    clayPagination = new ClayPagination({
      spritemap: icon,
      total: 7,
      page: 5,
    });

    expect(clayPagination).toMatchSnapshot();
  });

  // eslint-disable-next-line
  it('shound render a pagination with active last page and disabled angle-right', () => {
    clayPagination = new ClayPagination({
      page: 7,
      spritemap: icon,
      total: 7,
    });

    expect(clayPagination).toMatchSnapshot();
  });

  // eslint-disable-next-line
  it('should render a pagination with all pages visible, ellipse range for 10 and active page 7', () => {
    clayPagination = new ClayPagination({
      ellipseRange: 10,
      page: 7,
      spritemap: icon,
      total: 10,
    });

    expect(clayPagination).toMatchSnapshot();
  });

  it('should render a pagination with dropdown and change url links', () => {
    clayPagination = new ClayPagination({
      baseHref: '?page=',
      ellipseRange: 10,
      page: 7,
      spritemap: icon,
      total: 12,
    });

    expect(clayPagination).toMatchSnapshot();
  });

  it('should render a pagination with dropdown and offset', () => {
    clayPagination = new ClayPagination({
      offset: 780,
      page: 7,
      spritemap: icon,
      total: 800,
    });

    expect(clayPagination).toMatchSnapshot();
  });

  // eslint-disable-next-line
  it('should calculate exactly how many items in the dropdown with random numbers', () => {
    const getRandomInt = (min = 200, max = 10000) => {
      min = Math.ceil(min);
      max = Math.floor(max);
      return Math.floor(Math.random() * (max - min)) + min;
    };

    for (let i = 0; i < 3; i++) {
      const getNumberToRandom = getRandomInt();

      clayPagination = new ClayPagination({
        offset: 200,
        spritemap: icon,
        total: getNumberToRandom,
      });

      const itemsToDropdown =
        getNumberToRandom -
        clayPagination.offset +
        1 -
        clayPagination.ellipseRange +
        3 -
        1;

      expect(clayPagination._rangeEllipse().length).toEqual(itemsToDropdown);
    }
  });
});
