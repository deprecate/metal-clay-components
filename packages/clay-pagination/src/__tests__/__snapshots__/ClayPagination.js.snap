// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayPagination should render a pagination markup 1`] = `
<ul class="pagination">
  <li class="page-item disabled">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="icons.svg#angle-left"></use>
      </svg>
    </a>
  </li>
  <li class="page-item active">
    <a class=" page-link" href="#1">1</a>
  </li>
  <li class="page-item disabled">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="icons.svg#angle-right"></use>
      </svg>
    </a>
  </li>
</ul>
`;

exports[`ClayPagination should render a pagination with 4 pages visible and 3 pages in dropdown 1`] = `
<ul class="pagination">
  <li class="page-item disabled">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="icons.svg#angle-left"></use>
      </svg>
    </a>
  </li>
  <li class="page-item active">
    <a class=" page-link" href="#1">1</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#2">2</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#3">3</a>
  </li>
  <li class="page-item">
    <div class="dropdown">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn page-link btn-link">...</button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class="dropdown-item" href="#4">4</a>
          </li>
          <li>
            <a class="dropdown-item" href="#5">5</a>
          </li>
          <li>
            <a class="dropdown-item" href="#6">6</a>
          </li>
        </ul>
      </div>
    </div>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#7">7</a>
  </li>
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="icons.svg#angle-right"></use>
      </svg>
    </a>
  </li>
</ul>
`;

exports[`ClayPagination should render a pagination with 4 pages visible and 3 pages in dropdown with page active in dropdown 1`] = `
<ul class="pagination">
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="icons.svg#angle-left"></use>
      </svg>
    </a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#1">1</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#2">2</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#3">3</a>
  </li>
  <li class="page-item">
    <div class="dropdown">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn page-link btn-link">...</button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class="dropdown-item" href="#4">4</a>
          </li>
          <li>
            <a class="dropdown-item active" href="#5">5</a>
          </li>
          <li>
            <a class="dropdown-item" href="#6">6</a>
          </li>
        </ul>
      </div>
    </div>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#7">7</a>
  </li>
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="icons.svg#angle-right"></use>
      </svg>
    </a>
  </li>
</ul>
`;

exports[`ClayPagination should render a pagination with all pages visible, ellipse range for 10 and active page 7 1`] = `
<ul class="pagination">
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="icons.svg#angle-left"></use>
      </svg>
    </a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#1">1</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#2">2</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#3">3</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#4">4</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#5">5</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#6">6</a>
  </li>
  <li class="page-item active">
    <a class=" page-link" href="#7">7</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#8">8</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#9">9</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#10">10</a>
  </li>
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="icons.svg#angle-right"></use>
      </svg>
    </a>
  </li>
</ul>
`;

exports[`ClayPagination should render a pagination with dropdown and change url links 1`] = `
<ul class="pagination">
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="icons.svg#angle-left"></use>
      </svg>
    </a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="?page=1">1</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="?page=2">2</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="?page=3">3</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="?page=4">4</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="?page=5">5</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="?page=6">6</a>
  </li>
  <li class="page-item active">
    <a class=" page-link" href="?page=7">7</a>
  </li>
  <li class="page-item">
    <div class="dropdown">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn page-link btn-link">...</button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class="dropdown-item" href="?page=8">8</a>
          </li>
          <li>
            <a class="dropdown-item" href="?page=9">9</a>
          </li>
          <li>
            <a class="dropdown-item" href="?page=10">10</a>
          </li>
          <li>
            <a class="dropdown-item" href="?page=11">11</a>
          </li>
        </ul>
      </div>
    </div>
  </li>
  <li class="page-item">
    <a class=" page-link" href="?page=12">12</a>
  </li>
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="icons.svg#angle-right"></use>
      </svg>
    </a>
  </li>
</ul>
`;

exports[`ClayPagination should render a pagination with dropdown and offset 1`] = `
<ul class="pagination">
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="icons.svg#angle-left"></use>
      </svg>
    </a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#780">780</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#781">781</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#782">782</a>
  </li>
  <li class="page-item">
    <div class="dropdown">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn page-link btn-link">...</button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class="dropdown-item" href="#783">783</a>
          </li>
          <li>
            <a class="dropdown-item" href="#784">784</a>
          </li>
          <li>
            <a class="dropdown-item" href="#785">785</a>
          </li>
          <li>
            <a class="dropdown-item" href="#786">786</a>
          </li>
          <li>
            <a class="dropdown-item" href="#787">787</a>
          </li>
          <li>
            <a class="dropdown-item" href="#788">788</a>
          </li>
          <li>
            <a class="dropdown-item" href="#789">789</a>
          </li>
          <li>
            <a class="dropdown-item" href="#790">790</a>
          </li>
          <li>
            <a class="dropdown-item" href="#791">791</a>
          </li>
          <li>
            <a class="dropdown-item" href="#792">792</a>
          </li>
          <li>
            <a class="dropdown-item" href="#793">793</a>
          </li>
          <li>
            <a class="dropdown-item" href="#794">794</a>
          </li>
          <li>
            <a class="dropdown-item" href="#795">795</a>
          </li>
          <li>
            <a class="dropdown-item" href="#796">796</a>
          </li>
          <li>
            <a class="dropdown-item" href="#797">797</a>
          </li>
          <li>
            <a class="dropdown-item" href="#798">798</a>
          </li>
          <li>
            <a class="dropdown-item" href="#799">799</a>
          </li>
        </ul>
      </div>
    </div>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#800">800</a>
  </li>
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="icons.svg#angle-right"></use>
      </svg>
    </a>
  </li>
</ul>
`;

exports[`ClayPagination should render a pagination with total of the pages visible 1`] = `
<ul class="pagination">
  <li class="page-item disabled">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="icons.svg#angle-left"></use>
      </svg>
    </a>
  </li>
  <li class="page-item active">
    <a class=" page-link" href="#1">1</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#2">2</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#3">3</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#4">4</a>
  </li>
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="icons.svg#angle-right"></use>
      </svg>
    </a>
  </li>
</ul>
`;

exports[`ClayPagination shound render a pagination with active last page and disabled angle-right 1`] = `
<ul class="pagination">
  <li class="page-item">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="icons.svg#angle-left"></use>
      </svg>
    </a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#1">1</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#2">2</a>
  </li>
  <li class="page-item">
    <a class=" page-link" href="#3">3</a>
  </li>
  <li class="page-item">
    <div class="dropdown">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn page-link btn-link">...</button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class="dropdown-item" href="#4">4</a>
          </li>
          <li>
            <a class="dropdown-item" href="#5">5</a>
          </li>
          <li>
            <a class="dropdown-item" href="#6">6</a>
          </li>
        </ul>
      </div>
    </div>
  </li>
  <li class="page-item active">
    <a class=" page-link" href="#7">7</a>
  </li>
  <li class="page-item disabled">
    <a class=" page-link">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="icons.svg#angle-right"></use>
      </svg>
    </a>
  </li>
</ul>
`;
